cpp/include/cudf/detail/gather.cuh

```diff
@@ -656,34 +657,49 @@ std::unique_ptr<table> gather(table_view const& source_table,
                               rmm::cuda_stream_view stream,
                               rmm::mr::device_memory_resource* mr)
 {
-  std::vector<std::unique_ptr<column>> destination_columns;
-
-  // TODO: Could be beneficial to use streams internally here
-
-  for (auto const& source_column : source_table) {
-    // The data gather for n columns will be put on the first n streams
-    destination_columns.push_back(
-      cudf::type_dispatcher<dispatch_storage_type>(source_column.type(),
-                                                   column_gatherer{},
-                                                   source_column,
-                                                   gather_map_begin,
-                                                   gather_map_end,
-                                                   bounds_policy == out_of_bounds_policy::NULLIFY,
-                                                   stream,
-                                                   mr));
+  auto const num_columns = source_table.num_columns();
+  auto result            = std::vector<std::unique_ptr<column>>(num_columns);
+
+  // The data gather for n columns will be executed over n streams. If there is
+  // only a single column, the fork/join overhead should be avoided.
+  auto streams = std::vector<rmm::cuda_stream_view>{};
+  if (num_columns > 1) {
+    streams = cudf::detail::fork_streams(stream, num_columns);
```

Comment:
I am curious how this works for per-thread default stream. For spark, we build cuDF with PTDS. Will `streams` will be have number-of-columns vector of the PTDS stream?

Reply:
The `stream` passed in would be the PTDS stream. Then a stream pool (for that thread) would be created, the work would be executed across that stream pool, and then the `join` step would insert events for all the elements of `streams` to be synchronized with `stream` before new work on `stream` (the PTDS stream in Spark's case) would be runnable.